### ISSUE TITLE
Different contrastive loss from the paper

### ISSUE THREAD

**Issue by Yuliang-Zou** [tfzhou/ContrastiveSeg#6]
Hi, thanks for the great work! 

However, I found that the contrastive loss [here](https://github.com/tfzhou/ContrastiveSeg/blob/main/lib/loss/loss_contrast.py#L121) seems to be different from Eqn. (3) in the paper. According to the paper, the denominator should be 1 specific positive pair plus all the negative pairs. But in the current implementation, the denominator will be all the positive pairs (excluding identical matching) plus all the negative pairs.

**Comment by tfzhou** [comment]
Hi, thanks for your interest. In short, we only compute the sum of `neg_logits` but not `exp_logits` .

More specifically, in [L121](https://github.com/tfzhou/ContrastiveSeg/blob/9a381b9799c16d81e18d8f9f25ab509b93fb56de/lib/loss/loss_contrast.py#L121), `exp_logits` has a shape of `NxM` and `neg_logits` has a shape of `Nx1`, where each element in `exp_logits` denotes the logit of each sample pair, while each element in `neg_logits` denotes the sum of logits of all negative pairs of one anchor, _i.e._, `neg_logits = neg_logits.sum(1, keepdim=True)`.  In this way, their sum `exp_logits + neg_logits` actually adds each anchor's sum of negative logits to each specific sample pair.

Please let me know if you have further concerns.

**Comment by weiaicunzai** [comment]
> Hi, thanks for your interest. In short, we only compute the sum of `neg_logits` but not `exp_logits` .
> 
> More specifically, in [L121](https://github.com/tfzhou/ContrastiveSeg/blob/9a381b9799c16d81e18d8f9f25ab509b93fb56de/lib/loss/loss_contrast.py#L121), `exp_logits` has a shape of `NxM` and `neg_logits` has a shape of `Nx1`, where each element in `exp_logits` denotes the logit of each sample pair, while each element in `neg_logits` denotes the sum of logits of all negative pairs of one anchor, _i.e._, `neg_logits = neg_logits.sum(1, keepdim=True)`. In this way, their sum `exp_logits + neg_logits` actually adds each anchor's sum of negative logits to each specific sample pair.
> 
> Please let me know if you have further concerns.

Thanks for your reply. However, I am still confused with line (L121)[https://github.com/tfzhou/ContrastiveSeg/blob/9a381b9799c16d81e18d8f9f25ab509b93fb56de/lib/loss/loss_contrast.py#L121].
``` python
#why use:
log_prob = logits - torch.log(exp_logits + neg_logits)
# instead of:
log_prob = exp_logits  - torch.log(exp_logits + neg_logits)
```
In Eq (3): I think ``logits `` here should be the exp(i*i_+ /t) in the numerator， which should be  ``exp_logits ``.
Thanks for you time.

**Comment by crisz94** [comment]
> > Hi, thanks for your interest. In short, we only compute the sum of `neg_logits` but not `exp_logits` .
> > More specifically, in [L121](https://github.com/tfzhou/ContrastiveSeg/blob/9a381b9799c16d81e18d8f9f25ab509b93fb56de/lib/loss/loss_contrast.py#L121), `exp_logits` has a shape of `NxM` and `neg_logits` has a shape of `Nx1`, where each element in `exp_logits` denotes the logit of each sample pair, while each element in `neg_logits` denotes the sum of logits of all negative pairs of one anchor, _i.e._, `neg_logits = neg_logits.sum(1, keepdim=True)`. In this way, their sum `exp_logits + neg_logits` actually adds each anchor's sum of negative logits to each specific sample pair.
> > Please let me know if you have further concerns.
> 
> Thanks for your reply. However, I am still confused with line (L121)[https://github.com/tfzhou/ContrastiveSeg/blob/9a381b9799c16d81e18d8f9f25ab509b93fb56de/lib/loss/loss_contrast.py#L121].
> 
> ```python
> #why use:
> log_prob = logits - torch.log(exp_logits + neg_logits)
> # instead of:
> log_prob = exp_logits  - torch.log(exp_logits + neg_logits)
> ```
> 
> In Eq (3): I think `logits ` here should be the exp(i*i_+ /t) in the numerator， which should be `exp_logits `. Thanks for you time.

log exp(a) = a
